File: extension/core/src/main/java/io/holunda/camunda/bpm/data/adapter/map/MapReadAdapterLockedExternalTask.java
package io.holunda.camunda.bpm.data.adapter.map;

import io.holunda.camunda.bpm.data.adapter.ReadAdapter;
import io.holunda.camunda.bpm.data.adapter.WrongVariableTypeException;
import org.camunda.bpm.engine.externaltask.LockedExternalTask;

import java.util.Collections;
import java.util.Map;
import java.util.Optional;

/**
 * Read adapter for external task
 *
 * @param <K> type of key.
 * @param <V> type of value.
 */
@SuppressWarnings("unchecked")
public class MapReadAdapterLockedExternalTask<K, V> implements ReadAdapter<Map<K, V>> {

  private final LockedExternalTask lockedExternalTask;
  private final String variableName;
  private Class<K> keyClazz;
  private Class<V> valueClazz;

  public MapReadAdapterLockedExternalTask(LockedExternalTask lockedExternalTask, String variableName, Class<K> keyClazz, Class<V> valueClazz) {
    this.lockedExternalTask = lockedExternalTask;
    this.variableName = variableName;
    this.keyClazz = keyClazz;
    this.valueClazz = valueClazz;
  }

  @Override
  public Map<K, V> get() {
    return null;
  }

  @Override
  @SuppressWarnings("java:S3655")
  public Optional<Map<K,V>> getOptional() {
    return Optional.ofNullable(getOrNull(getValue()));
  }

  @Override
  public Map<K,V> getLocal() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public Optional<Map<K,V>> getLocalOptional() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public Map<K,V> getOrDefault(Map<K,V> defaultValue) {
    return getOptional().orElse(defaultValue);
  }

  @Override
  public Map<K,V> getLocalOrDefault(Map<K,V> defaultValue) {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public Map<K,V> getOrNull() {
    return getOrNull(getValue());
  }

  @Override
  public Map<K,V> getLocalOrNull() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @SuppressWarnings("Duplicates")
  private Map<K,V> getOrNull(Object value) {
    if (value == null) {
      return null;
    }

    if (Map.class.isAssignableFrom(value.getClass())) {
      Map<?, ?> valueAsMap = (Map<?, ?>) value;
      if (valueAsMap.isEmpty()) {
        return Collections.emptyMap();
      } else {
        Map.Entry<?, ?> entry = valueAsMap.entrySet().iterator().next();
        if (keyClazz.isAssignableFrom(entry.getKey().getClass()) && valueClazz.isAssignableFrom(entry.getValue().getClass())) {
          return (Map<K, V>) valueAsMap;
        } else {
          throw new WrongVariableTypeException("Error reading " + variableName + ": Wrong map type detected, expected Map<"
            + keyClazz.getName() + "," + valueClazz.getName()
            + ", but was not found in " + valueAsMap);
        }
      }
    }

    throw new WrongVariableTypeException("Error reading " + variableName + ": Couldn't read value of type Map from " + value);
  }

  private Object getValue() {
    return Optional.ofNullable(lockedExternalTask.getVariables()).map(it -> it.get(variableName)).get();
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/adapter/map/MapReadAdapterLockedExternalTask.kt
package io.holunda.camunda.bpm.data.adapter.map

import org.camunda.bpm.engine.externaltask.LockedExternalTask
import org.camunda.bpm.engine.variable.Variables
import java.util.*

/**
 * Read adapter for external task
 *
 * @param [K] type of key.
 * @param [V] type of value.
 * @param lockedExternalTask external task.
 * @param variableName name of the variable.
 */
class MapReadAdapterLockedExternalTask<K, V>(
  private val lockedExternalTask: LockedExternalTask,
  variableName: String,
  keyClazz: Class<K>,
  valueClazz: Class<V>
) : AbstractMapReadWriteAdapter<K, V>(variableName, keyClazz, valueClazz) {

  private val value: Any?
    get() = Optional.ofNullable(lockedExternalTask.variables)
      .orElse(Variables.createVariables())[variableName]

  override fun getOptional(): Optional<Map<K, V>> {
    return Optional.ofNullable(
      getOrNull(
        value
      )
    )
  }

  override fun getLocal(): Map<K, V> {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun getLocalOptional(): Optional<Map<K, V>> {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun getLocalOrDefault(defaultValue: Map<K, V>): Map<K, V> {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun getLocalOrNull(): Map<K, V> {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun remove() {
    throw UnsupportedOperationException("Can't remove a variable on an external task")
  }

  override fun removeLocal() {
    throw UnsupportedOperationException("Can't remove a local variable on an external task")
  }

  override fun set(value: Map<K, V>, isTransient: Boolean) {
    throw UnsupportedOperationException("Can't set a variable on an external task")
  }

  override fun setLocal(value: Map<K, V>, isTransient: Boolean) {
    throw UnsupportedOperationException("Can't set a local variable on an external task")
  }

}
